### PR TITLE
Gracefully Handle Missing Permissions

### LINKED ISSUE #4: Permission errors should not cause job failure
# 🐛 Bug Report

A permission error should be reported, but not failed upon.  See: 

https://github.com/cisagov/action-lineage/runs/772741162?check_suite_focus=true

## Any helpful log output

Paste the results here:

```
2020-06-15T13:16:56.6114233Z INFO Checking: cisagov/cool-system
2020-06-15T13:16:56.8016847Z INFO Lineage configuration found for cisagov/cool-system
2020-06-15T13:16:56.8017470Z INFO Cloning repository: https://github.com/cisagov/cool-system.git
2020-06-15T13:16:57.3166350Z INFO ✅ success
2020-06-15T13:16:57.3167776Z INFO Processing lineage: skeleton
2020-06-15T13:16:57.3168742Z INFO Upstream: https://github.com/cisagov/skeleton-generic.git HEAD
2020-06-15T13:16:57.3169069Z INFO Attempting to switch to branch: lineage/skeleton
2020-06-15T13:16:57.3194201Z INFO ✅ (error ok) return code: 128
2020-06-15T13:16:57.3194784Z INFO Branch did not exist.  Creating: lineage/skeleton from local develop
2020-06-15T13:16:57.3195117Z INFO Creating branch lineage/skeleton from develop
2020-06-15T13:16:57.3219662Z INFO ✅ success
2020-06-15T13:16:57.3221732Z INFO Switching to lineage/skeleton
2020-06-15T13:16:57.3443533Z INFO ✅ success
2020-06-15T13:16:57.3443679Z INFO Pull request branch is new: True
2020-06-15T13:16:57.6403691Z INFO Fetching https://github.com/cisagov/skeleton-generic.git HEAD
2020-06-15T13:16:57.6404068Z INFO ✅ success
2020-06-15T13:16:57.6429427Z INFO ✅ success
2020-06-15T13:16:57.6462396Z INFO ✅ success
2020-06-15T13:16:57.6462555Z INFO Attempting merge of fetched changes.
2020-06-15T13:16:57.7075221Z INFO ✅ success
2020-06-15T13:16:57.7075723Z INFO Remove any incoming modifications to .github/lineage.yml
2020-06-15T13:16:57.7292196Z INFO ✅ success
2020-06-15T13:16:57.7514505Z INFO ✅ success
2020-06-15T13:16:57.7515575Z INFO Committing merge.
2020-06-15T13:16:57.7783416Z INFO ✅ success
2020-06-15T13:16:57.7787494Z INFO Assigning credentials for push.
2020-06-15T13:16:57.7818077Z INFO ✅ success
2020-06-15T13:16:57.7818718Z INFO Pushing lineage/skeleton to remote.
2020-06-15T13:16:57.9309812Z CRITICAL remote: Permission to cisagov/cool-system.git denied to cisagovbot.
2020-06-15T13:16:57.9310313Z fatal: unable to access 'https://github.com/cisagov/cool-system.git/': The requested URL returned error: 403
2020-06-15T13:16:57.9310435Z 
2020-06-15T13:16:57.9310700Z CRITICAL ❌ ERROR! return code: 128
2020-06-15T13:16:57.9380220Z Traceback (most recent call last):
2020-06-15T13:16:57.9380557Z   File "/src/lineage/entrypoint.py", line 379, in <module>
2020-06-15T13:16:57.9380684Z     sys.exit(main())
2020-06-15T13:16:57.9384292Z   File "/src/lineage/entrypoint.py", line 336, in main
2020-06-15T13:16:57.9384606Z     push(repo, pr_branch_name, "git", access_token)
2020-06-15T13:16:57.9384728Z   File "/src/lineage/entrypoint.py", line 190, in push
2020-06-15T13:16:57.9385276Z     run([GIT, "push", "--set-upstream", "origin", branch_name], cwd=repo.full_name)
2020-06-15T13:16:57.9385422Z   File "/src/lineage/entrypoint.py", line 67, in run
2020-06-15T13:16:57.9385553Z     raise Exception("Subprocess was expected to exit with 0.")
2020-06-15T13:16:57.9385682Z Exception: Subprocess was expected to exit with 0.
2020-06-15T13:16:58.3991507Z Cleaning up orphan processes
```

### PR DESCRIPTION
**Pull request by mcdonnnj**
# 

## 🗣 Description

This PR adds functionality to gracefully handle a previously fatal case of missing permissions. This should resolve #4 . It also adds a status badge to the README for the `lineage_scan` workflow.

## 💭 Motivation and Context

We have run into the issue where a repository does not have the correct permissions configured, and this causes the Lineage workflow to fail. If unnoticed, it can go on for an extended period of time (the latest was four weeks). When this happens, the workflow will only process repositories until it hits the problem repository. Anything in the search results after that will not be updated.

The new badge allows an at-a-glance check for the last `lineage_scan` workflow run.

## 🧪 Testing

Automated tests pass.

## 🚥 Types of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/lineage/entrypoint.py [hunk 1]
```diff
@@ -180,14 +180,25 @@ def merge(repo: Repository.Repository, github_actor: str) -> Tuple[bool, List[st
...
     """Push changes to remote."""
-    parts: ParseResult = urlparse(repo.clone_url)
-    cred_url = parts._replace(netloc=f"{username}:{password}@{parts.netloc}").geturl()
-    logging.info("Assigning credentials for push.")
-    run([GIT, "remote", "set-url", "origin", cred_url], cwd=repo.full_name)
-    logging.info(f"Pushing {branch_name} to remote.")
-    run([GIT, "push", "--set-upstream", "origin", branch_name], cwd=repo.full_name)
+    if not repo.permissions.push:
+        logging.warning(
+            f"⚠️ ERROR! Missing 'push' permission on '{repo.owner.login}/{repo.name}'"
```
Comment (dav3r) [inline src/lineage/entrypoint.py]: There's a bit of a mismatch here between us logging a `warning`, but calling it `ERROR!` in the message itself.

Comment (mcdonnnj) [reply src/lineage/entrypoint.py]: That's copied from:
https://github.com/cisagov/action-lineage/blob/9d4ec4d62ce3ed9e6b219bd4f9bc0e7cc6dac3f7/src/lineage/entrypoint.py#L61-L63

Comment (jsf9k) [reply src/lineage/entrypoint.py]: I agree with @dav3r.  It should be consistent.  I'd change it anywhere else it appears as well.

File: src/lineage/entrypoint.py [hunk 2]
```diff
@@ -180,14 +180,25 @@ def merge(repo: Repository.Repository, github_actor: str) -> Tuple[bool, List[st
...
+        )
+        return False
+    else:
+        parts: ParseResult = urlparse(repo.clone_url)
+        cred_url = parts._replace(
+            netloc=f"{username}:{password}@{parts.netloc}"
+        ).geturl()
+        logging.info("Assigning credentials for push.")
+        run([GIT, "remote", "set-url", "origin", cred_url], cwd=repo.full_name)
```
Comment (jsf9k) [inline src/lineage/entrypoint.py:198]: At some point we should replace these git commands with [GitPython](https://pypi.org/project/GitPython/).

**Review by jsf9k — CHANGES_REQUESTED** [review]
I'd like those `logging.warning("ERROR!")` messages to be made consistent.  Otherwise this looks good to me.

File: src/lineage/entrypoint.py (hunk 2 above)
Comment (jsf9k) [reply src/lineage/entrypoint.py:198]: See #12, which I just created.

Comment (mcdonnnj) [reply src/lineage/entrypoint.py:198]: @felddy Do you remember why you didn't leverage that when you first wrote this?

File: src/lineage/entrypoint.py (hunk 1 above)
Comment (mcdonnnj) [reply src/lineage/entrypoint.py]: Please see 4eb817f

File: src/lineage/entrypoint.py (hunk 2 above)
Comment (felddy) [reply src/lineage/entrypoint.py:198]: Probably didn't know about it yet.  Not a technical reason.

File: src/lineage/entrypoint.py [hunk 3]
```diff
@@ -333,42 +344,47 @@ def main() -> int:
...
+                        body = pystache.render(clean_template, template_data)
+                        create_pull_request(
+                            repo, pr_branch_name, local_branch, title, body, draft=False
+                        )
                 else:
-                    title = f"Lineage pull request for: {lineage_id}"
-                    body = pystache.render(clean_template, template_data)
-                    create_pull_request(
-                        repo, pr_branch_name, local_branch, title, body, draft=False
+                    logging.info(
```
Comment (felddy) [inline src/lineage/entrypoint.py]: Should we make this a warning?

File: src/lineage/entrypoint.py [hunk 4]
```diff
@@ -333,42 +344,47 @@ def main() -> int:
...
+                        "Not creating a new pull request since the branch already existed."
                     )
             else:
                 logging.info(
```
Comment (felddy) [inline src/lineage/entrypoint.py]: Warning?

**Review by felddy — APPROVED** [review]
Looks good. Thanks for fixing my omissions.   See my two ideas about warnings.

File: src/lineage/entrypoint.py (hunk 3 above)
Comment (jsf9k) [reply src/lineage/entrypoint.py]: Let's make it a warning, but make sure the warning string starts with "ERROR!'

**Review by jsf9k — APPROVED** [review]
Strong work!

File: src/lineage/entrypoint.py (hunk 1 above)
Comment (dav3r) [reply src/lineage/entrypoint.py]: Thanks @mcdonnnj 👍

**Review by dav3r — APPROVED** [review]
Nice fixes in here 🚀

**Pull request merged**